### PR TITLE
[Test] Login relay

### PR DESCRIPTION
**Pull request by NOOMA-42**
## Summary

`login.test.ts` and implement the Backend tests for login and sign-up

## Linked Issue

#34 

## Details

* Test relay, login case 1 2 (wallet, server)
* Implement the Backend tests for login and sign-up
Beside test login relay:
* Add document 

## Impacted Areas

* `login.test.ts`
* `TransactionManager.ts`: [0e1a6be](https://github.com/social-tw/social-tw-website/pull/47/commits/0e1a6be422621c0caefad22ef5c4de4b42d31240)
* `post.test.ts`: we branched out from test/post-relay
* `UserService.ts`: encapsulation [b172b2a](https://github.com/social-tw/social-tw-website/pull/47/commits/b172b2a1cfd685b76f5d04ce426bf8c4e5692592)
* `relay/ts.config`: add sourceMap option for debug
 
Add following doc:
* `Contributing.md`
* `Doc Folder`: including local development, system overview, etc. [68cab7c](https://github.com/social-tw/social-tw-website/pull/47/commits/68cab7c665bcf8957f7ece6cd36abe0e8ea18d23)

## Tests

N/A

## Possible Impacts

1. `TransactionManager` throw error instead of console logging
2. Abstract away UserService from `/api/user`

## Visual Materials

N/A

## Verification Steps

run `yarn relay test` in root folder

## Todo

Break down the progress of the PR for everyone to see what else you intend to include in this PR.

-   [X] The backend should have a /signup API.
-   [X]  It should accept [sign up proof](https://developer.unirep.io/docs/circuits-api/circuits#signup-proof) and the router should verify it
-   [X]  It should accepct twitter Oauth2 information and record the used twitter userid.
-   [X]  It should return tx success or fail.

## Checklist

-   [X] All new and existing tests pass
-   [X] I have updated the documentation (if applicable)
-   [X] My changes do not introduce new security vulnerabilities

### CODE REVIEW

**Comment by FoodChain1028** [comment]
please modify the PR to the correct name 😂

**Comment by NOOMA-42** [comment]
@cc03668 Is one of our todo 'It should accepct Github Oauth information and record the used Github Account.' wrongly typed? Should it be 'It should accepct twitter Oauth2 information and record the used twitter userid.' instead?

**Comment by NOOMA-42** [comment]
@ryanycw @jason-ntu  
1. I was wondering how to get this lint to pass.
2. we don't have a test command for this branch frontend, can we ignore it? Or what should we do?
@ryanycw 
3. Do you have a situation where the contract doesn't go through?

**Comment by ryanycw** [comment]
I believe we only need 

> @ryanycw @jason-ntu
> 
> 1. I was wondering how to get this lint to pass.
We could do `yarn lint:fix` to get it passed. We could use husky to solve this.

> 2. we don't have a test command for this branch frontend, can we ignore it? Or what should we do?
>    @ryanycw
We could ignore it for now. We should be good as long as we get the circuit, relay, and lint pass!

> 3. Do you have a situation where the contract doesn't go through?
Yes, in our branch, the contract tests aren't included yet.

**Comment by cc03668** [comment]
@NOOMA-42 correct! it should be Twitter instead of GitHub. Sorry, it's a C&P from Vivian's spec based on previous unirep social.

**Review by ryanycw — APPROVED** [review]
Thank you for the great work! I only left some comments for future improvement. I believe these tests are valid. And we could decide if we want to make the changes in this PR or open a issue as future work.

File: packages/relay/test/login.test.ts [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+const mockState = 'state'
+const mockCode = 'testCode'
+const mockCode2 = 'testCode2'
+const mockUserId = '123456'
+const mockUserId2 = '654321'
+const wrongCode = 'wrong-code'
+const token = btoa(`${TWITTER_CLIENT_ID}:${TWITTER_CLIENT_KEY}`)
+
+describe('LOGIN /login', () => {
+    let snapshot: any
```
Comment (chentihe) [inline packages/relay/test/login.test.ts:29]: Is snapshot necessary for this test? It seems that some tests are rely on others. You will revert to the snapshot to have a clean storage before executing the next test. If so, you should move revert statement to ```afterEach```

Comment (chentihe) [reply packages/relay/test/login.test.ts:29]: You're right, I comment the snapshot and the error occurred on post test. I will resolve this conversation and approve the review.

**Review by chentihe — APPROVED** [review]

**Pull request merged**